In [ ]:
# <redacted> <redacted>
# https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
# Commit ID: 9c91722

In [40]:
# Imports section
import pandas as pd
from sklearn import model_selection as ms
from sklearn import linear_model as lm
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures as pf
from sklearn import svm




## Part 1. Loading the dataset

In [41]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv("science_data_large.csv")
print ("First 15 Rows of Dataset:")
print (data.head(15))
print ("\nStatistics:")
data.agg(
    {
        "Temperature °C": ["min", "max", "median", "skew"],
        "Mols KCL": ["min", "max", "median", "mean"],
        "Size nm^3": ["min", "max", "median", "mean"],
    }
)




First 15 Rows of Dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Statistics:


,Temperature °C,Mols KCL,Size nm^3
min,1.0,1.00,1.611429e+01
max,1000.0,1000.00,1.972127e+06
median,500.5,459.50,3.827182e+05
skew,0.0,NaN,NaN
mean,NaN,471.53,5.086111e+05


## Part 2. Splitting the dataset

In [42]:
# Take the pandas dataset and split it into our features (X) and label (y)
x = data[["Temperature °C", "Mols KCL"]]
y = data["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = ms.train_test_split(x,y, test_size = 0.1, train_size = 0.9, random_state=42)


## Part 3. Perform a Linear Regression

In [43]:
# Use sklearn to train a model on the training set
model = lm.LinearRegression()
reg = model.fit(x_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[400, 700]]
print (reg.predict(sample))

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print (reg.score(x_test, y_test))

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print (f"Intercept = ", reg.intercept_)
print (f"Coefficients = ", reg.coef_)

[659953.63230542]
0.8552472077276095
Intercept =  -409391.47958340764
Coefficients =  [ 866.14641337 1032.69506649]


c:\Users\baldw\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Given that the point (403, 694) from our training data has an actual size of about 577000, the estimate of the point (400,700) having a size of 660000 seems to be fairly reasonable. 
The score of this model is about 0.85, meaning the predictions are all faily close to actual data. R^2 values range from 0-1 and the closer they are to one the more reliable the data is. A value of .85 means it is fairly accurate but can still be improved.



Model Equation = $h(x) = 1032.7x_2 + 866.1x_1 - 409391.5$

$x_1$ = Temperature in degrees Celcius

$x_2$ = Moles of Potassium chloride


## Part 4. Use Cross Validation

In [44]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = ms.cross_val_score(model, x, y, cv=10)
print (f"Scores: {scores}")
print (f"Average R^2 value = {scores.mean()}")
# Report on their finding and their significance

Scores: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Average R^2 value = 0.8552450341984701


Findings: These values are similar to the R^2 value I got in part 3, which means that the lines were all slightly different from each other. looking at the scores given here i can assume that the score given in the first trial of 0.85 was one of the lines that most closely matches the data.

## Part 5. Using Polynomial Regression

In [45]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = pf(degree=2)
poly_x_train = poly.fit_transform(x_train)
poly_x_test = poly.fit_transform(x_test)
polyreg = lm.LinearRegression().fit(poly_x_train,y_train)

print (polyreg.score(poly_x_test, y_test))
print (f"Coefficients = {polyreg.coef_}")
print (f"Intercept = {polyreg.intercept_}")
# Report on the metrics and output the resultant equation as you did in Part 3.

1.0
Coefficients = [ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]
Intercept = 2.0479375962167978e-05


These findings show that the data is perfectly matched by the equation below. An R^2 value of 1 means that the line lays right on top of every single data point in the sample. This explains why the straight line was not able to reach a score higher than 0.9.



Equation: $h(x) = -0.0000204 + 12x_1 + (-1.27*10^{-7})x_2 + (1.28*10^{-11})x_1^2 + 2x_2^2 + 0.0285(x_1 * x_2)$

$x_1$ = Temperature in degrees Celcius

$x_2$ = Moles of Potassium chloride

